projects/storefrontlib/src/cms-components/misc/qualtrics/qualtrics-loader.service.ts

```diff
@@ -1,78 +1,131 @@
+  protected initialize() {
+    this.qsi$.subscribe(() => this.run());
+  }
 
-    this.winRef.document
-      .getElementsByTagName('head')[0]
-      .appendChild(qualtricsScript);
+  /**
+   * Evaluates the Qualtrics project code for the application.
+   *
+   * In order to reload the evaulation in Qualtrics, the API requires to unload the API before
+   * running it again. We don't do this by default, but offer a flag to conditionally unload the API.
+   */
+  protected run(reload = false): void {
+    if (!this.qsiApi?.API) {
+      if (isDevMode()) {
+        console.log('The QSI api is not available');
+      }
+      return;
+    }
 
-    this.winRef.document.getElementsByTagName('head')[0].appendChild(idScript);
-  }
+    if (reload) {
+      // Removes any currently displaying creatives
+      this.qsiApi.API.unload();
+    }
 
-  private isQualtricsConfigured(): boolean {
-    return (
-      Boolean(this.config.qualtrics) && Boolean(this.config.qualtrics.projectId)
-    );
+    // Starts the intercept code evaluation right after loading the Site Intercept
+    // code for any defined intercepts or creatives
+    this.qsiApi.API.load().done(this.qsiApi.API.run());
   }
 
-  load(): Observable<boolean> {
-    return this.qualtricsLoaded$.pipe(
-      filter((loaded) => loaded),
-      switchMap(() => {
-        const qsi = this.winRef.nativeWindow['QSI'];
-        return this.isDataLoaded().pipe(
-          distinctUntilChanged(),
-          tap((dataLoaded) => {
-            if (dataLoaded) {
-              qsi.API.unload();
-              qsi.API.load().done(qsi.API.run());
-            }
-          })
-        );
-      })
-    );
+  /**
+   * Adds the deployment script to the DOM. If the deployment script is not
+   *
+   * The script will not be added twice if it was loaded before. In that case, we _run_
+   * the Qualtrics API straight away.
+   */
+  addScript(scriptSource: string): void {
+    if (this.hasScript(scriptSource)) {
+      this.run(true);
+    } else {
+      const script: HTMLScriptElement = this.renderer.createElement('script');
+      script.type = 'text/javascript';
+      script.defer = true;
+      script.src = scriptSource;
+      this.renderer.appendChild(this.winRef.document.body, script);
+    }
   }
 
   /**
-   * This logic exist in order to let the client(s) add their own logic to wait for any kind of page data
-   * If client(s) does not extend this service to override this implementation, it returns true
-   * Return false otherwise.
+   * This logic exist in order to let the client(s) add their own logic to wait for any kind of page data.
+   * You can observe any data in this method.
+   *
+   * Defaults to true.
    */
   protected isDataLoaded(): Observable<boolean> {
     return of(true);
   }
+
+  /**
+   * Indicates if the script is already added to the DOM.
+   */
+  protected hasScript(source?: string): boolean {
+    return !!this.winRef.document?.querySelector(`script[src="${source}"]`);
```

Comment:
Is the `?` really needed in `this.winRef.document?`?

Reply:
Good catch, Domino [actually support querySelector](https://github.com/fgnass/domino/blob/master/README.md) .